src/infra/core/src/repos/metadata_chain_visitor_impl.rs

```diff
@@ -0,0 +1,340 @@
+// Copyright Kamu Data, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use chrono::{DateTime, Utc};
+use kamu_core::{
+    AppendError,
+    AppendValidationError,
+    BlockNotFoundError,
+    OffsetsNotSequentialError,
+    SequenceIntegrityError,
+};
+use opendatafabric::{DataSlice, IntoDataStreamBlock, MetadataBlock, MetadataEvent, Multihash};
+
+use crate::{
+    invalid_event,
+    BoxedVisitors,
+    Decision,
+    MetadataBlockTypeFlags,
+    MetadataBlockWithOptionalHashRef,
+    MetadataChainVisitor,
+};
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+pub struct ValidateSeedBlockOrderVisitor {}
+
+impl MetadataChainVisitor for ValidateSeedBlockOrderVisitor {
+    fn visit(
+        &mut self,
+        (block, _): MetadataBlockWithOptionalHashRef,
+    ) -> Result<Decision, AppendError> {
+        match block.event {
+            MetadataEvent::Seed(_) if block.prev_block_hash.is_some() => {
+                Err(AppendValidationError::AppendingSeedBlockToNonEmptyChain.into())
+            }
+            MetadataEvent::Seed(_) => Ok(Decision::Stop),
+            _ if block.prev_block_hash.is_none() => {
+                Err(AppendValidationError::FirstBlockMustBeSeed.into())
+            }
+            _ => Ok(Decision::Stop),
+        }
+    }
+}
+
+///////////////////////////////////////////////////////////////////////////////
+
+#[derive(Default)]
+pub struct ValidatePrevBlockExistsVisitor {
+    next_prev_block_hash: Option<Multihash>,
+}
+
+impl MetadataChainVisitor for ValidatePrevBlockExistsVisitor {
```

Comment:
Perhaps this simply shouldn't be a visitor. This is something we can very easily check in `append` with a lot fewer lines of code.

Reply:
I think we can leave it in -- this logic is just for Validator.
1) it requests a block from the past
2) it validates :)